js/private/js_info.bzl

```diff
@@ -3,53 +3,73 @@
 JsInfo = provider(
     doc = "Encapsulates information provided by rules in rules_js and derivative rule sets",
     fields = {
+        "target": "The label of target that created this JsInfo",
+        "bin_dir": "The root corresponding to bin directory for the target",
+        "sources": "A depset of source files produced by the target",
         "types": "A depset of typings files produced by the target",
+        "transitive_sources": "A depset of source files produced by the target and the target's transitive deps",
+        "transitive_types": "A depset of declaration files produced by the target and the target's transitive deps",
         "npm_sources": "A depset of files in npm package dependencies of the target and the target's transitive deps",
         "npm_package_store_infos": "A depset of NpmPackageStoreInfo providers from non-dev npm dependencies of the target and the target's transitive dependencies to use as direct dependencies when linking downstream npm_package targets with npm_link_package",
-        "sources": "A depset of source files produced by the target",
-        "transitive_types": "A depset of declaration files produced by the target and the target's transitive deps",
-        "transitive_sources": "A depset of source files produced by the target and the target's transitive deps",
     },
 )
 
 def js_info(
-        types = depset(),
-        npm_sources = depset(),
-        npm_package_store_infos = depset(),
+        target,
+        bin_dir,
         sources = depset(),
+        types = depset(),
+        transitive_sources = depset(),
         transitive_types = depset(),
-        transitive_sources = depset()):
+        npm_sources = depset(),
+        npm_package_store_infos = depset()):
     """Construct a JsInfo.
 
     Args:
+        target: See JsInfo documentation
+        bin_dir: See JsInfo documentation
+        sources: See JsInfo documentation
         types: See JsInfo documentation
+        transitive_sources: See JsInfo documentation
+        transitive_types: See JsInfo documentation
         npm_sources: See JsInfo documentation
         npm_package_store_infos: See JsInfo documentation
-        sources: See JsInfo documentation
-        transitive_types: See JsInfo documentation
-        transitive_sources: See JsInfo documentation
 
     Returns:
         A JsInfo provider
     """
+    if type(target) != "Label":
+        msg = "Expected target to be a Label but got {}".format(type(target))
```

Comment:
Why add the `msg` var everywhere?

Reply:
If you don't `msg` then Bazel shows the template string in the error message which is confusing. Using `msg` makes it show us as `fail(msg)` on the callstack when it is hit